File: exercise-solutions/shoppers-solution.scala
import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.SparkConf

import org.apache.spark.storage.StorageLevel

import org.apache.spark.sql.SQLContext






object ShoppersApp extends Application {

    // Create a Spark SparkContext
    val conf = new SparkConf().setAppName("Shoppers Application")
                              .set("spark.executor.memory", "1g")
                              .set("spark.storage.memoryFraction","0.3")
    val sc = new SparkContext(conf)


    /* History class: history of incentive offered a customer and information about the behavioral response to the offer

    customer_id - A unique id representing a customer
    chain - An integer representing a store chain
    offer_id - An id representing a certain offer
    market - An id representing a geographical region
    repeattrips - The number of times the customer made a repeat purchase
    repeater - A boolean, equal to repeattrips > 0
    offerdate - The date a customer received the offer)
    */ 
    case class History(customer_id: String, chain: Int, offer_id: String, market: String, repeattrips: Int, repeater: Boolean, offerdate: String) 

    /* Transaction 
    customer_id - see above
    chain - see above
    dept - An aggregate grouping of the Category (e.g. water)
    category - The product category (e.g. sparkling water)
    company - An id of the company that sells the item
    brand - An id of the brand to which the item belongs
    date - The date of purchase
    productsize - The amount of the product purchase (e.g. 16 oz of water)
    productmeasure - The units of the product purchase (e.g. ounces)
    purchasequantity - The number of units purchased
    purchaseamount - The dollar amount of the purchase
    */ 
    case class Transaction(customer_id: String, chain: Int, dept: String, category: String, company: String, brand: String,
                             date: String, productsize: Double, productmeasure: String, purchasequantity: Int, purchaseamount: Double)


    /* Offer class: 
    offer_id - see above
    category - see above
    quantity - The number of units one must purchase to get the discount
    company - see above
    offervalue - The dollar value of the offer
    brand - see above
    */ 
    case class Offer(offer_id: String, category: String, quantity: Int, company: String, offervalue: Double, brand: String)


    // Load the CSV files as texts


    // RDD of case objects of type History
    val history = sc.textFile("../data/shoppers/history") map { line => 
        val h = line.split(',')
        def isRepeater(s: String) = s == 't'
        History(h(0), h(1).toInt, h(2), h(3), h(4).toInt, isRepeater(h(5)), h(6))
    } 
    
    // RDD of case objects of type Offer
    val offers = sc.textFile("../data/shoppers/offers") map { line => 
        val o = line.split(',')
        Offer(o(0), o(1), o(2).toInt, o(3), o(4).toDouble, o(5))
    } 

    


    // ----------------------------------- YOUR CODE FOLLOWS HERE ------------------------------------- 


    // Q1: Create RDDs of key-value pairs for 'history' and 'offers' and join them into histOffers Pair RDD
    val histOffers = (history map (h => (h.offer_id, h))) join (offers map (o => (o.offer_id, o))) 
    histOffers.cache()


    // Q2: Find the popular offers that resulted in 5 most repeattrips and return a collection of 3-tuple (customer_id, offer_id, repeattrips)
    val top5Offers: Array[(String, String, Int, Double)] = histOffers.groupBy{case (_, (h,_)) => h.repeattrips}.
                                                                      sortByKey(false).take(5).flatMap(t => t._2).
                                                                      map{case (id, (h,o)) => (h.customer_id, id, h.repeattrips, o.offervalue)}


    // Then print these Top 5 Offers
    top5Offers.foreach { case (customer, offer, repeattrips, value) => 
        println("Customer #" + customer + " saved " + value + " dollars on offer #" + offer + " and later made repeteated purchases " + repeattrips + " times")
    }

    /*  Should print: 
            Customer #3465135195 saved 0.75 dollars on offer #1197502 and later made repeteated purchases 2124 times
            Customer #3450535153 saved 0.75 dollars on offer #1197502 and later made repeteated purchases 1418 times
            Customer #4427711419 saved 1.0 dollars on offer #1203052 and later made repeteated purchases 549 times
            Customer #4176323168 saved 1.0 dollars on offer #1203052 and later made repeteated purchases 104 times
            Customer #239276354 saved 0.75 dollars on offer #1197502 and later made repeteated purchases 87 times
    */ 


    // Extract top 5 customers 
    val top5Customers = top5Offers.map{ case (c: String, _, _, _) => c}

    // Optional: find the maximum number of repeattrips generated by offers with a 5$ discount value
    val maxRepeatsFromFiveBucks = {
        val offersFiveBucks = histOffers.filter {case (_, (h,o)) => o.offervalue == 3}
                                        .groupBy {case (id, (h, o)) => h.repeattrips}
                                        .sortByKey(false).first()
        offersFiveBucks._1
    }
    println("\nCoupons with 5 dollar value yield a maximum of only " + maxRepeatsFromFiveBucks + " repeated purchases")



    // Let us now look into the purchase history of these top 5 Customers prior to their offered incentive

    // The dataset containing the transactions of these top 5 customers are given to you as textfiles in topTransactions folder 
    // Let us first load into RDD of case objects of type Transaction
    
    val inputTrans = sc.textFile("../data/shoppers/topTransactions")   

    val transactions = (inputTrans map { line => 
        val t = line.split(',')
        Transaction(t(0), t(1).toInt, t(2), t(3), t(4), t(5), t(6), t(7).toDouble, t(8), t(9).toInt, t(10).toDouble)
    })


    // Q4: Compute the average spending Per purchase trip (assuming one trip per day) of these 5 customers

    // To compute this one, we will divide it into multiple tasks: 
    val avgSpendingPerTrip = {

        // Map the transactions by Customer and Date, i.e. the key consists of a pair (customer_id, date)
        val transKeyedByCustomerAndDate = transactions map {t => ((t.customer_id, t.date), t)}

        // Compute the total spending per trip per customer using combineByKey or reduceByKey
        val spendingPerTrip = transKeyedByCustomerAndDate.combineByKey(
          (t) => (t.purchaseamount), 
          (acc: Double, t) => (acc + t.purchaseamount),
          (a1: Double, a2: Double) => (a1 + a2))

        // Transform spendingPerTrip into an RDD of pairs (customer_id, totalSpendingValue) for each trip 
        val spendingsPerCustomer = spendingPerTrip map {case (trip: (String, String), value: Double) => (trip._1, value)}

        // Compute the average spending per trip for each customer
        val averageSpendings = spendingsPerCustomer.mapValues(s => (s,1))
                                                     .reduceByKey{(a: (Double, Int), b: (Double, Int)) => (a._1 + b._1, a._2 + b._2)}
                                                     .mapValues(t => t._1 / t._2)

        averageSpendings.collect()  // collect result from RDD averageSpendings
    }   

    avgSpendingPerTrip.foreach{case (customer, avg) => println("Customer #" + customer + " : " + avg)}
        

    // Bonus Question: For each customer, compute the average spending Per Purchase trip only for items that have  
    // the same Category as the one for which they were offered a coupon, e.g. same category in transactions and history. 


}

File: exercise-solutions/shoppers-solution.py
#!/usr/bin/env python
import csv
import sys
import itertools
import StringIO

from operator import add
from os.path import join, isfile, dirname

from pyspark import SparkContext, SparkConf

# ------------------------------
#  Some local parser functions
#-------------------------------
def loadOffer(line):
  input = StringIO.StringIO(line)
  reader = csv.DictReader(input, fieldnames = ["offer_id", "category","quantity","company","offervalue","brand"])    
  return reader.next()

def convertTypeOffer(o):
  o['quantity'] = int(o['quantity']) 
  o['offervalue'] = float(o['offervalue'])
  return o

def loadHistory(line):
  input = StringIO.StringIO(line)
  reader = csv.DictReader(input, fieldnames = ["customer_id", "chain","offer_id","market","repeattrips","repeater", "offerdate"])
  return reader.next()

def convertTypeHistory(h):
  h['repeattrips'] = int(h['repeattrips'])
  h['chain'] = int(h['chain'])
  h['repeater'] = {'t': True}.get(h['repeater'], False)  
  return h

def loadTransaction(line):
  input = StringIO.StringIO(line)
  reader = csv.DictReader(input, fieldnames = ["customer_id", "chain","dept","category","company","brand","date","productsize","productmeasure","purchasequantity","purchaseamount"])    
  return reader.next()

def convertTypeTransaction(t):
  t['chain'] = int(t['chain'])
  t['productsize'] = float(t['productsize'])
  t['purchasequantity'] = int(t['purchasequantity'])
  t['purchaseamount'] = float(t['purchaseamount'])
  return t

#-----------------------------------------


if __name__ == "__main__":


  


  conf = SparkConf().setAppName("ShoppersApp").set("spark.executor.memory","1g")
  sc = SparkContext(conf = conf)

  # Load the files into RDDs of History, Offer and Transaction dictionnaries
  history = sc.textFile("../data/shoppers/history").map(loadHistory).map(convertTypeHistory)
  offers = sc.textFile("../data/shoppers/offers").map(loadOffer).map(convertTypeOffer)


  # ----------------------------------- YOUR CODE FOLLOWS HERE ------------------------------------- 

  # Q1: Create RDDs of key-value pairs for 'history' and 'offers' and join them into histOffers Pair RDD
  def byOffer_Id(x): 
    return (x['offer_id'], x) 
  histOffers = (history.map(byOffer_Id)).join(offers.map(byOffer_Id))

  # Q2: Find the popular offers that resulted in 5 most repeattrips and return a collection of 3-tuple (customer_id, offer_id, repeattrips)
  top5Offers = histOffers.groupBy(lambda elmt: elmt[1][0]['repeattrips']) \
                         .sortByKey(False).flatMap(lambda t: t[1]) \
                         .map(lambda elmt: (elmt[1][0]['customer_id'], elmt[0], elmt[1][0]['repeattrips'], elmt[1][1]['offervalue'])) \
                         .take(5)

  # Then print these Top 5 Offers
  for x in top5Offers:
    print("Customer #" + str(x[0]) + " saved " + str(x[3]) + " dollars on offer #" + str(x[1]) + " and later made repeteated purchases " + str(x[2]) + " times")

    # Should print: 
    #         Customer #3465135195 saved 0.75 dollars on offer #1197502 and later made repeteated purchases 2124 times
    #         Customer #3450535153 saved 0.75 dollars on offer #1197502 and later made repeteated purchases 1418 times
    #         Customer #4427711419 saved 1.0 dollars on offer #1203052 and later made repeteated purchases 549 times
    #         Customer #4176323168 saved 1.0 dollars on offer #1203052 and later made repeteated purchases 104 times
    #         Customer #239276354 saved 0.75 dollars on offer #1197502 and later made repeteated purchases 87 times

  # Extract top 5 customers
  top5Customers = [x[0] for x in top5Offers]

  # // Q3: find the maximum number of repeattrips generated by offers with a 5$ discount value
  offersFiveBucks = histOffers.filter(lambda elmt: elmt[1][1] ['offervalue'] == 3) \
                              .groupBy (lambda elmt: elmt[1][0]['repeattrips']) \
                              .sortByKey(False).first()
                              
  maxRepeatsFromFiveBucks = offersFiveBucks[0]
  print("\nCoupons with 5 dollar value yield a maximum of only " + str(maxRepeatsFromFiveBucks) + " repeated purchases")

  # Let us now look into the purchase history of these top 5 Customers prior to their offered incentive

  # The dataset containing the transactions of these top 5 customers are given to you as textfiles in topTransactions folder 
  # Let us first load into RDD 

  transactions = sc.textFile("../data/shoppers/topTransactions", 100).map(loadTransaction).map(convertTypeTransaction)


  # RDD of case objects of type Transaction

  # Q4: Compute the average spending Per purchase trip (assuming one trip per day) of these 5 customers

  # Map the transactions by Customer and Date, i.e. the key consists of a pair (customer_id, date)
  transKeyedByCustomerAndDate = transactions.map(lambda t : ((t['customer_id'], t['date']), t))

  # Compute the total spending per trip per customer using combineByKey or reduceByKey
  spendingPerTrip = transKeyedByCustomerAndDate.combineByKey(
          (lambda t : t['purchaseamount']), 
          (lambda acc, t : acc + t['purchaseamount']),
          (lambda a1, a2 : a1 + a2))

  # Transform spendingPerTrip into an RDD of pairs (customer_id, totalSpendingValue) for each trip 
  spendingsPerCustomer = spendingPerTrip.map(lambda x:  (x[0][0], x[1]))

  # Compute the average spending per trip for each customer
  avgSpendingPerTrip = spendingsPerCustomer.mapValues(lambda s: (s,1)) \
                                           .reduceByKey(lambda a, b : (a[0] + b[0], a[1] + b[1])) \
                                           .mapValues(lambda t: t[0] / t[1])  \
                                           .collect()
                                           
                                           
  # Print the average spending per trip for each customer
  print(" ")
  for (customer, avg) in avgSpendingPerTrip:
   print("Customer #" + customer + " : " + str(avg))

  # Bonus Question: For each customer, compute the average spending Per Purchase trip only for items that have  
  # the same Category as the one for which they were offered a coupon, e.g. same category in transactions and history. 
  
  # clean up
  sc.stop()

